add option submits_with to button

pardeyke opened the pull request:
as discussed here #781

@lcreid what do you think?

Also should there be a screenshot in den README?

<img width="121" height="42" alt="Screenshot 2025-11-27 at 11 01 09" src="https://github.com/user-attachments/assets/71d1cc75-351f-427e-be74-bca3ae98974c" />

pardeyke committed:
add option submits_with to button

pardeyke committed:
use spinner by default

pardeyke committed:
undo

lcreid commented on lib/bootstrap_form/inputs/submit.rb:
```diff
@@ -4,6 +4,10 @@ module BootstrapForm
     module Submit
       def button(value=nil, options={}, &)
+        if options.key? :submits_with
+          options[:data] = { turbo_submits_with: setup_turbo_submit(options[:submits_with]) }
```
I think this should `merge` with `options[;data]`, in case the user needs to specify other `data-` attributes.

lcreid commented on test/bootstrap_other_components_test.rb:
```diff
@@ -152,6 +152,22 @@ class BootstrapOtherComponentsTest < ActionView::TestCase
+  test "regular button has turbo-submits-with deafault spinner" do
+    expected = <<~HTML
+      <button class="btn btn-secondary" data-turbo-submits-with="<div class=&quot;spinner-border d-block mx-auto&quot; role=&quot;status&quot; style=&quot;--bs-spinner-width: 1lh; --bs-spinner-height: 1lh;&quot;></div>" name="button" type="submit">Submit with Spinner</button>
+    HTML
```
I don't think the expected HTML should be escaped.

lcreid commented on README.md:
````diff
@@ -1065,6 +1065,20 @@ will be rendered as
+<button class="btn btn-secondary" data-turbo-submits-with="<div class="spinner-border d-block mx-auto"; role="status" style="--bs-spinner-width: 1lh; --bs-spinner-height: 1lh;"></div>" name="button" type="submit">Save</button>
+```
+
+Use `submits_with: :spinner` to render a default bootstrap spinner or pass your own HTML.
````
It would be nice to show an example where the user provides the HTML.

lcreid commented on README.md:
````diff
@@ -1065,6 +1065,20 @@ will be rendered as
+```
+
+Use `submits_with: :spinner` to render a default bootstrap spinner or pass your own HTML.
+This only works on `f.button` or `f.primary` not on `f.submit` and forces `render_as_button: true` on `f.primary`.
````
I would move this line up to near the top of this section, because some of us don't read the whole section before we start coding. ;-)

lcreid reviewed (commented):
Thanks for this PR. This is a lot like what I was describing in your previous PR.

I really like the idea of making it easier to put up a spinner. I will confess that I hadn't even considered how to do this (I don't do enough front-end work these days). At the same time, I think we should think a lot about how to add new features, so we don't end up committing ourselves to supporting something that turns out not to be what most users need.

With that in mind, I wonder what you about trying to reduce the scope of this PR a bit? What I'm thinking is that instead of a new parameter for `f.button` or `f.primary`, we just provide a helper that only generates this HTML for a spinner, and let the user code the `data-turbo-submits-with` attribute themself. Something like:

```
f.button("Save", data: { turbo_submits_with: f.spinner })
```

This would allow us in the future to make the default spinner more configurable (e.g. with keyword parameters with defaults), without breaking existing applications that were happy with the default spinner.

What do you think?

Also, you're right that we should have an example and screenshot for the documentation. Please add that if you feel you a can figure out how from the existing examples. If run into trouble, feel free to ask me for help.

pardeyke commented:
I like your idea to let the user code the data attribute themselves.

I can't figure out where the f.spinner should live. Can you point me to file and how to set it up?

pardeyke committed:
-let users put in the data attribute themselves
-update readme
-todo: add f.spinner

pardeyke committed:
updated tests

lcreid commented:
> I can't figure out where the f.spinner should live. Can you point me to file and how to set it up?

Good question. I suggest putting it in `lib/bootstrap_form/helpers/bootstrap.rb`. There are some methods in that file that are meant to be used by developers using `bootstrap_form`, like `f.spinner`.

pardeyke committed:
Merge branch 'main' of https://github.com/bootstrap-ruby/bootstrap_form into turbo_submits_with_option

pardeyke committed:
-moving spinner to helper
-handeling [:data] = nil
-fix tests (use html entities here because f.spinner goes in a data-turbo-submits-with attribute)

pardeyke commented:
Sorry it took some time... December is always a mess

pardeyke commented:
@lcreid What do you think? Are we ready to merge? 😃
